### PR TITLE
Deployment improvements + OIDC

### PR DESCRIPTION
**Pull request by ividito**
Tracked by: #292 

## What

- Add OIDC and secrets management pattern (comparable to https://github.com/NASA-IMPACT/veda-stac-ingestor/pull/48 , modified slightly for this project's architecture)
- Convert deployment action to use OIDC to fetch env vars and deploy resources (the only Github secret needed now are OIDC secrets)

**Note:** This PR doesn't fully adopt OIDC in our default deployment action. I want to get this merged so I can try some test runs on a non-local runner first.

## Why

- Making it easier to track deployed resources
- Simplifying Github actions flow (currently dependant on administering an account with access to resources - now this repo has its own role with limited permissions)

## How to test

At least one manual deployment is required to initialize OIDC resources and secrets.

Using `act`, the `deploy_oidc.yml` action can be tested using the following.

```
act -W .github/workflows/deploy_oidc.yml --secret-file .secrets --env ACT=true -P ubuntu-latest=nektos/act-environments-ubuntu:18.04 --input DEPLOY_ENV=isayah -v
```

My `.secrets` file contains the following:

```
OIDC_ROLE_ARN=arn:aws:iam::853558080719:role/<fill this in>
GITHUB_TOKEN=<needs action permissions - https://docs.github.com/en/rest/overview/permissions-required-for-github-apps?apiVersion=2022-11-28>
AWS_ACCESS_KEY_ID=<any account with access to OIDC role and/or secrets>
AWS_SECRET_ACCESS_KEY=X
```

To test OIDC functionality, I manually extended a deployed role to grant access to my AWS account. By default, only a action being run from this repository will be able to use OIDC (thus the conditional step in `deploy_oidc`).

### CODE REVIEW

File: .github/workflows/deploy_oidc.yml [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      - name: Install node and related deps
+        uses: actions/setup-node@v3
+        with:
+          node-version: 16.19.1
+
+      - name: Install AWS CDK
+        run: npm install -g aws-cdk
+
+      - name: Configure awscli (local runner)
+        if: ${{ env.ACT }} # OIDC doesn't work when using ACT, as AWS user accounts don't have access to the OIDC role by default
```
Comment (moradology) [inline .github/workflows/deploy_oidc.yml:41]: I found this a bit confusing but perhaps I don't understand something relevant about role assumption: isn't it the case that roles to be assumed depend on `trust` relationships which will be related to user or account ID? This isn't a correction so much as a curiosity about what I am missing

File: .github/workflows/deploy_oidc.yml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+      - name: Configure awscli (non-local runner)
+        if: ${{ !env.ACT }}
+        uses: aws-actions/configure-aws-credentials@v2
+        with:
+          role-to-assume: ${{ secrets.OIDC_ROLE_ARN }}
+          aws-region: us-west-2
+
+      - name: Get environment configuration from AWS Secrets Manager
+        run: |
+          aws secretsmanager get-secret-value --secret-id data-pipelines-env-secret-${{ github.event.inputs.DEPLOY_ENV }} --query SecretString --output text --region us-west-2 | jq -r "to_entries|map(\"\(.key)=\(.value|tostring)\")|.[]" >> $GITHUB_ENV
```
Comment (moradology) [inline .github/workflows/deploy_oidc.yml:57]: Surprised `jq` is available out of the box. Useful to know

File: deploy/app.py [hunk 3]
```diff
@@ -1,17 +1,38 @@
...
+app = App()
 
-env_details = core.Environment(
+config = cfg.Config(_env_file=".env")
+
+git_sha = subprocess.check_output(["git", "rev-parse", "HEAD"]).decode().strip()
+try:
+    git_tag = subprocess.check_output(["git", "describe", "--tags"]).decode().strip()
+except subprocess.CalledProcessError:
+    git_tag = "no-tag"
```
Comment (moradology) [inline deploy/app.py:24]: Very nice. Definitely superior to the sha-only strategy I've used in the past for this kind of thing

**Review by moradology — APPROVED** [review]
A few comments/questions but this looks good +1

File: .github/workflows/deploy_oidc.yml (hunk 1 above)
Comment (ividito) [reply .github/workflows/deploy_oidc.yml:41]: As-is, the role requires a token/OIDC auth from this Github repository, so (as an example) my AWS user account running locally, without that auth from Github, cannot assume the role unless that condition is modified.

**Pull request merged**